std/regex/package.d

```diff
@@ -356,13 +356,13 @@ public alias StaticRegex = Regex;
 
     Throws: `RegexException` if there were any errors during compilation.
 +/
-@trusted public auto regex(S)(S[] patterns, const(char)[] flags="")
+@trusted public auto regex(S : C[], C)(const S[] patterns, const(char)[] flags="")
 if (isSomeString!(S))
 {
     import std.array : appender;
     import std.functional : memoize;
     enum cacheSize = 8; //TODO: invent nice interface to control regex caching
-    S pat;
+    const(C)[] pat;
```

Comment:
If the only reason to introduce `C` is this, maybe `ElemType` would be better ?

Reply:
I don't think so (maybe I'm missing something?), because:
1. `regexImpl` supports only string types (not ranges).
2. `ElemType` (I assume you mean `std.range.primitives.ElementType`) will return `dchar`, given `T = string`, or `T = wstring`. In case we're called with such narrow strings, we would need to re-encode them to dchar[], which would be the type of `pat`.
3. `const`-ness. Changing the parameter from `S[] patterns` to `const S[] patterns` is not strictly necessary here (but it is for `regexImpl`), but by doing this I'm saving on distinct template instances. Before for `S = char[]`, `S = const(char)[]` and `S = const(char[])` (ditto for `immutable`) we would get different template instances. Now there's only one.
So, with this change in mind, when `regex(const S[] patterns, ..)` is called with a `char[][]`, `patterns` is deduced to be `const(char[][])`, while `S` becomes `char[]` and we get the following error: 
    ```
    Error: cannot implicitly convert expression patterns[0] of type const(char[]) to char[]
    ```
    (Because in the line `pat = patterns[0]`, `pat` is `S = char[]` and `patterns[0]` is `const(char[])`.)
     Additionally, I can't define `pat` as `const S` as then I won't be able to assign to it.
    So what I need is the char type, so I can define `pat` as tail-`const`.
4. Using the template specialization syntax `S : C[], C` is much shorter and readable then `ElementEncodingType` (from `std.range.primitives`). It's probably also slightly faster to compile.